internal/pkg/agent/application/upgrade/upgrade_test.go

```diff
@@ -740,3 +743,148 @@ func TestIsSameVersion(t *testing.T) {
+			states:              []details.State{details.StateWatching},
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             50 * time.Millisecond,
+			wantErr:             assert.NoError,
+		},
+		{
+			name:                "Sad path: watcher is never starting",
+			states:              []details.State{details.StateReplacing},
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             50 * time.Millisecond,
+			wantErr:             wantErrWatcherNotStarted,
+		},
+		{
+			name: "Runaround path: marker is jumping around and landing on watching",
+			states: []details.State{
+				details.StateRequested,
+				details.StateScheduled,
+				details.StateDownloading,
+				details.StateExtracting,
+				details.StateReplacing,
+				details.StateRestarting,
+				details.StateWatching,
+			},
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             100 * time.Millisecond,
+			wantErr:             assert.NoError,
+		},
+		{
+			name:                "Timeout: marker is never created",
+			states:              nil,
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             50 * time.Millisecond,
+			wantErr:             wantErrWatcherNotStarted,
+		},
+		{
+			name: "Timeout2: state doesn't get there in time",
+			states: []details.State{
+				details.StateRequested,
+				details.StateScheduled,
+				details.StateDownloading,
+				details.StateExtracting,
+				details.StateReplacing,
+				details.StateRestarting,
+				details.StateWatching,
+			},
+
+			stateChangeInterval: 5 * time.Millisecond,
+			timeout:             20 * time.Millisecond,
+			wantErr:             wantErrWatcherNotStarted,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			deadline, ok := t.Deadline()
+			if !ok {
+				deadline = time.Now().Add(5 * time.Second)
+			}
+			ctx, cancel := context.WithDeadline(context.TODO(), deadline)
+			defer cancel()
+
+			tmpDir := t.TempDir()
+			updMarkerFilePath := filepath.Join(tmpDir, markerFilename)
+
+			if len(tt.states) > 0 {
+				initialState := tt.states[0]
+				writeState(t, updMarkerFilePath, initialState)
+			}
+
+			wg := new(sync.WaitGroup)
+
+			var furtherStates []details.State
+			if len(tt.states) > 1 {
+				// we have more states to produce
+				furtherStates = tt.states[1:]
+
+				wg.Add(1)
+
+				go func() {
+					defer wg.Done()
+					tick := time.NewTicker(tt.stateChangeInterval)
+					defer tick.Stop()
+					for _, state := range furtherStates {
+						select {
+						case <-ctx.Done():
+							return
+						case <-tick.C:
+							writeState(t, updMarkerFilePath, state)
+						}
+					}
+
+				}()
+			}
+
+			log, _ := logger.NewTesting(tt.name)
+
+			tt.wantErr(t, waitForWatcher(ctx, log, updMarkerFilePath, tt.timeout), fmt.Sprintf("waitForWatcher %s, %v, %s, %s)", updMarkerFilePath, tt.states, tt.stateChangeInterval, tt.timeout))
+
+			// cancel context
+			cancel()
+
+			//wait for goroutines to finish
```

Comment:
please, run go fmt on save

```suggestion
			// wait for goroutines to finish
```